### ISSUE TITLE
CodecException during read request with Id conflict between path and resource TLV.

### ISSUE THREAD

**Issue by Gopishankarp** [eclipse/leshan#460]
CodecException occurred while reading a single instance resource value (lifetime) of LWM2M server object.
Path : [1/0/1].
**Exception message:**
```
org.eclipse.leshan.core.request.exception.InvalidResponseException: Unable to decode response payload of request [ReadRequest [path=/10245/0/4 format=ContentFormat [name=TLV, code=11542]]] from client
...
Caused by: org.eclipse.leshan.core.node.codec.CodecException: Id conflict between path [/1/0/1] and resource TLV [0]
	at org.eclipse.leshan.core.node.codec.tlv.LwM2mNodeTlvDecoder.parseTlv(LwM2mNodeTlvDecoder.java:134) ~[LWM2M_Mediator-1.0.M3.jar:1.0]
```

**Response from the client:**
<img width="470" alt="tlv_response" src="https://user-images.githubusercontent.com/22880479/35106558-8db29aaa-fc94-11e7-94d8-5f2ef05ffa3b.PNG">

**Full stack trace of exception**
[TLVCodecException.log](https://github.com/eclipse/leshan/files/1643430/TLVCodecException.log)

Is it server issue or invalid reponse from client

**Comment by sbernard31** [comment]
I'm a bit confused, you talk about the Lifetime(id 1) resource in Server(id 1) object, so path /1/0/1.
And you stack trace talk about object with id 10245?

**Comment by sbernard31** [comment]
Could you provide the TLV payload in hexa too?

**Comment by Gopishankarp** [comment]
@sbernard31 Same exception/problem occurred for both 10245/0/4 and 1/0/1 . In fact any read request.

**Comment by sbernard31** [comment]
Again the stack trace let me think that the problem is at client side.
By the way, I just try to read 1/0/1 using Leshan Server Demo and Leshan Client Demo and it works for me.

Without more details I can not help :/

**Comment by Gopishankarp** [comment]
It is because in the TLV header identifier type is 00 which is object instance (00.. .... = Type of Identifier: Object Instance (0) ) when we doing a read request of resource. ( means -> 11 instead of 00 )

**Comment by sbernard31** [comment]
So we agree the issue was in the payload ?

**Comment by Gopishankarp** [comment]
I am asking if it is a issue with the payload? If yes , what is the reason for the exception. When the checked the class ( LwM2mNodeTlvDecoder.java) This exception when there is problem in the Type.
```
else if (tlvs.length == 1 && tlvs[0].getType() != TlvType.RESOURCE_INSTANCE) {
                if (path.isResource() && path.getResourceId() != tlvs[0].getIdentifier()) {
                    throw new CodecException("Id conflict between path [%s] and resource TLV [%s]", path,
                            tlvs[0].getIdentifier());
                }
```

**Comment by sbernard31** [comment]
**If you read a Resource(/X/X/X), we expect a TLV with type `resource`.**
This is not clearly defined in the spec but we never think that someone will send something else.
If OMA say that object instance TLV could be sent on resource request we can implement it but I need OMA explanation first. (you could ask [here](https://github.com/OpenMobileAlliance/OMA_LwM2M_for_Developers/issues))

Anyway, If you want to point us that the message is not really adapted... I agree I will change it.

**Comment by Gopishankarp** [comment]
I have created a issue in OMA seeking clarification.
https://github.com/OpenMobileAlliance/OMA_LwM2M_for_Developers/issues/276

**Comment by sbernard31** [comment]
(if you have the whole TLV payload in Hexa this could help)

**Comment by sbernard31** [comment]
Here is the payload decoded : 
```
[Tlv[
     type=OBJECT_INSTANCE, 
     children=[Tlv[
                type=RESOURCE_VALUE,
                children=null,
                value=[0, 0, -61, 80],
                identifier=1]
                ],
     value=null,
     identifier=0]
]
```
This is the payload for the Read Request on the resource /1/0/1, right ?

**Comment by Gopishankarp** [comment]
Yes. Below is the log from LWM2M Server.

> DEBUG o.e.l.c.n.c.DefaultLwM2mNodeDecoder - Decoding value for path /1/0/1 and format ContentFormat [name=TLV, code=11542]: [6, 0, -60, 1, 0, 0, -61, 80]02:46:59.594 [CoapServer#4]
 DEBUG o.e.l.s.c.impl.LwM2mResponseBuilder - Unable to decode response payload of request [ReadRequest [path=/1/0/1 format=ContentFormat [name=TLV, code=11542]]] from client [LTE-Relay_2102311HPF10F8000010] [payload:0600c4010000c350]

Is it problem with server ?


--

**Comment by sbernard31** [comment]
This confirm what I thought :

1. As I understand the spec : The main error is at client side, this payload is no correct ! Let's see what will be the answer of OpenMobileAlliance/OMA_LwM2M_for_Developers#276
2. At server side : The message error is not good and should be clearer.

Waiting if you really want to support this kind of "Bad client", you can implements your own Decoder.

see `LeshanServerBuilder.setDecoder()`

**Comment by Gopishankarp** [comment]
Thanks for your confirmation. 
Could you please tell what is the problem from the client side response. If possible what is the expected response .

**Comment by sbernard31** [comment]
As you read a Resource (the path contains 3 parts), we expect a TLV of type RESOURCE INSTANCE (01) or MULTIPLE RESOURCE (10).

So the payload should look like this : 
```
[Tlv[
           type=RESOURCE_VALUE,
           children=null,
           value=[0, 0, -61, 80],
           identifier=1]
]
```

**Comment by sbernard31** [comment]
As I can not obtain a clear answer from OMA ... (https://github.com/OpenMobileAlliance/OMA_LwM2M_for_Developers/issues/276)
I decided to support this in Leshan as we always flexible for TLV decoding.
So the PR about that is available : #514 

@Gopishankarp  Could you check if this behave better now ?

**Comment by sbernard31** [comment]
This should be fixed in master since #514 is integrated in master.
